### ISSUE TITLE
implement type-ahead for auto-completion

### ISSUE THREAD

**Issue by 257** [vim/vim#3433]
this is for having first candidate from pumvisible to be inserted in front of the cursor _tentatively_.

it really just comes down to the difference between autocompletion (we already have that) and typeahead as loosely define here: 
https://www.flextras.com/blog/index.cfm/2011/3/11/The-Difference-Between-AutoComplete-and-LookAhead

here is an animation of typeahead in action:
https://www.arungudelli.com/wp-content/uploads/2013/10/Twitter-Typeahead-js.gif

xcode, apparently, has this implemented already:
https://i.stack.imgur.com/k8PRF.png

**Comment by jamessan** [comment]
This is how completion already works.  With the default value for `'completeopt'`, the full text for the first match is inserted.

If you press `<C-e>` to cancel the completion, then whatever text you've explicitly typed remains, instead of what was inserted by the completion.

If you press `<C-y>` to accept the completion, it remains and the cursor is at the end of the text.

Otherwise, you can continue to type to refine the matches.

The only difference is that the animations you show leave the cursor at the end of the text that the user has typed and shows the completed text after the cursor (and faded).  There's no functional difference, just eye candy.

**Comment by caksoylar** [comment]
Would a different highlight group for the type-ahead part be useful in this case? Then colorscheme authors could define a lower contrast color for the part that is inserted with ins-completion but not yet accepted with `<C-y>`.

**Comment by brammool** [comment]
> Would a different highlight group for the type-ahead part be useful in
> this case? Then colorscheme authors could define a lower contrast
> color for the part that is inserted with ins-completion but not yet
> accepted with `<C-y>`.

Yeah, that would be possible, highlight the text added by the completion
with some color, such as grey.

It would look nice.  Might be a bit of work to implement...


-- 
"I simultaneously try to keep my head in the clouds and my feet on the
ground.  Sometimes it's a stretch, though."              -- Larry Wall

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
